### DISCUSSION TITLE
Is there a replacement for ImmediateRenderer in v11.0?

### DISCUSSION THREAD

**Discussion by FreddyDgh** [AvaloniaUI/Avalonia#10386] [Q&A] [answered]
@kekekeks I see in Avalonia v11.0 that `ImmediateRenderer` has been made internal in #10078.  Looks like it's been dropped in favor of the `CompositingRenderer`.  Previously, I had been rendering static content to a Skia canvas like so:

```csharp
IVisual visual;

using var context = new DrawingContext(DrawingContextHelper.WrapSkiaCanvas(canvas, SkiaPlatform.DefaultDpi));
ImmediateRenderer.Render(visual, context);
```

Is there an easy equivalent to render static content to a Skia canvas like this in v11.0, either via `CompositingRenderer` or some other mechanism?

**Comment by maxkatz6** [comment]
I don't think there is now.
You can draw over a wrapped skia canvas:
```
using var contextImpl = DrawingContextHelper.WrapSkiaCanvas(canvas, SkiaPlatform.DefaultDpi);
using var context = new DrawingContext(contextImpl);
context.Draw...
```
Or you can render existing controls to RenderTargetBitmap:
```
 using (var bitmap = new RenderTargetBitmap(pixelSize, dpiVector))
 {
    bitmap.Render(root);
    bitmap.Save(destination);
 }
```
But I don't see any way to combine RenderTargetBitmap with a custom drawing context.
CC @kekekeks

**Reply by FreddyDgh** [reply]
Thanks for the reply; it is helpful for planning.  Ultimately, I think it does make more sense to have the `ImmediateRenderer` be internal and expose the functionality through something like a `RenderTargetBitmap` that allows for a custom drawing context.  The Skia rendering functionality is nice to have for my scenario, but not strictly necessary.  Seems reasonable to put priorities elsewhere.

**Reply by maxkatz6** [reply]
Actually, I was wrong. It should be possible with this code.
Something like this should work
```csharp
private void DrawVisualToCanvas(SKCanvas canvas, Visual visual)
{
    using var contextImpl = DrawingContextHelper.WrapSkiaCanvas(canvas, SkiaPlatform.DefaultDpi);
    using var context = new DrawingContext(contextImpl);

    using var bitmap = new RenderTargetBitmap(new PixelSize(100, 100));
    bitmap.Render(visual);
    contextImpl.DrawBitmap(bitmap.PlatformImpl, 1,
        new Rect(0, 0, 100, 100),
        new Rect(0, 0, 100, 100));
}
```
Just need to adjust sizes, probably get it from the visual bounds.

**Reply by FreddyDgh** [reply]
The `DrawBitmap` method will draw the rendered UI to the canvas, but it is not entirely the same as rendering with the Skia drawing context.  For example, if you save the SkiaCanvas to a PDF file, you will not be able to search for text in the PDF file, because what gets saved to the PDF is not actually text - it's all one big image.

**Reply by maxkatz6** [reply, marked as answer]
For 11.1 (nightly currently) you can also use this method: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Skia/Avalonia.Skia/Helpers/DrawingContextHelper.cs#L14

**Reply by FreddyDgh** [reply]
Just got a chance to try this out.  It appears to work exactly how I wanted, so hopefully it will come to stable soon. Thanks for your efforts.

**Reply by maxkatz6** [reply]
It is already
